src/upgrade.c

```diff
@@ -164,12 +164,20 @@ static int upgrade_main(void *ctx, struct apk_database *db, struct apk_string_ar
 	if (uctx->self_upgrade_only)
 		return 0;
 
-	if (uctx->ignore) {
+	if (args->num) {
+		unsigned int flags = solver_flags, in_flags;
+		if (uctx->ignore) {
+			flags |= APK_SOLVERF_IGNORE_UPGRADE;
+			in_flags = 0;
+		} else {
+			solver_flags &= ~APK_SOLVERF_UPGRADE;
```

Comment:
I just reset the `upgrade` bit but it might maybe be more convenient to just put all the flags to 0 ?

Reply:
The if() rule to setup solver_flags should not be in the inner loop. It could be setup earlier, perhaps already in the options parsing, or by latest in the applet main function start.
What comes to the resetting the `solver_flags` variable it's probably better to reset it to zero - other flags might have side effects e.g. specifying the `_AVAILABLE` flag would cause re-installs in certain conditions.

Reply:
I'm not sure what you mean with 'inner loop'. I just check if there are any arguments provided then I update the flags accordingly.

I agree that this can be made earlier. The whole bloc might be moved before the `self-upgrade` and the `solver_flags` might be updated to always ignore self-upgrade if this is a named package upgrade.

Reply:
Seems I misread the patch somehow. Please disregard the 'inner loop' comment. But perhaps the options parsing could directly update solver_flags instead of uctx->ignore now that the test is based on number of arguments. We should be able to get rid of the whole ignore:1 bit.